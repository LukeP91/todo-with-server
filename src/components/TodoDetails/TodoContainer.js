import _ from 'lodash'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'
import { withStyles } from '@material-ui/core/styles'
import Paper from '@material-ui/core/Paper'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import Typography from '@material-ui/core/Typography'

import { fetchLabels } from '../../store/reducers/labels/actions'
import { fetchUsers } from '../../store/reducers/users/actions'
import { toggleTodo, removeTodo, editTodo } from '../../store/reducers/todos/actions'
import TodoDetails from './TodoDetails'

const styles = theme => ({
  paper: {
    padding: theme.spacing.unit * 2,
    textAlign: 'center',
    color: theme.palette.text.secondary,
  },
})

class TodoContainer extends Component {
  componentDidMount() {
    this.props.fetchUsers('http://localhost:3001/users', { method: 'GET' })
    this.props.fetchLabels('http://localhost:3001/labels', { method: 'GET' })
  }

  completeTodo = todo => {
    const { toggleTodo } = this.props

    toggleTodo(`http://localhost:3001/todos/${todo.id}`, {
      method: 'PATCH',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ completed: !todo.completed }),
    })
  }

  deleteTodo = id => {
    const { removeTodo, history } = this.props

    removeTodo(`http://localhost:3001/todos/${id}`, {
      method: 'DELETE',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
    })
    history.push({ pathname: '/' })
  }

  updateTodo = values => {
    const { editTodo, match } = this.props
    editTodo(`http://localhost:3001/todos/${match.params.id}`, {
      method: 'PATCH',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(values),
    })
  }

  render() {
    const { todos, classes, match, labels, users } = this.props
    const todo = _.find(todos, { id: parseInt(match.params.id, 10) })
    if (todo) {
      return (
        <TodoDetails
          completeTodo={this.completeTodo}
          deleteTodo={this.deleteTodo}
          labels={labels}
          todo={todo}
          updateTodo={this.updateTodo}
          users={users}
        />
      )
    }
    return (
      <Paper className={classes.paper}>
        <Typography align="left" gutterBottom variant="display1">
          Loading...
        </Typography>
      </Paper>
    )
  }
}

const mapStateToProps = state => ({
  todos: state.todos.todos,
  users: state.users.users,
  labels: state.labels.labels,
})

const mapDispatchToProps = {
  fetchLabels,
  fetchUsers,
  removeTodo,
  toggleTodo,
  editTodo,
}

TodoContainer.propTypes = {
  classes: PropTypes.object.isRequired,
  fetchLabels: PropTypes.func,
  fetchUsers: PropTypes.func,
  history: PropTypes.object,
  match: PropTypes.object.isRequired,
  labels: PropTypes.array,
  removeTodo: PropTypes.func,
  todos: PropTypes.array,
  toggleTodo: PropTypes.func,
  users: PropTypes.array,
  editTodo: PropTypes.func,
}

export default withRouter(
  withStyles(styles)(
    connect(
      mapStateToProps,
      mapDispatchToProps,
    )(TodoContainer),
  ),
)
