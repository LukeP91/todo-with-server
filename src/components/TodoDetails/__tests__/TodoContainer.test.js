import React from 'react'
import renderer from 'react-test-renderer'
import { MemoryRouter } from 'react-router-dom'

import { TodoContainer } from '../TodoContainer'

describe('TodoContainer', () => {
  const props = {
    classes: {},
    fetchLabels: jest.fn(),
    fetchUsers: jest.fn(),
    labels: [{ id: 1, value: 'test' }],
    match: { params: { id: 1 } },
    todos: [{ id: 1, title: 'some title' }],
    users: [{ id: 1, name: 'Joe Doe' }],
  }

  it('renders correctly and fetches users and labels on mount', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <TodoContainer {...props} />
        </MemoryRouter>,
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
    expect(props.fetchLabels).toBeCalledWith('http://localhost:3001/labels', { method: 'GET' })
    expect(props.fetchUsers).toBeCalledWith('http://localhost:3001/users', { method: 'GET' })
  })
})
