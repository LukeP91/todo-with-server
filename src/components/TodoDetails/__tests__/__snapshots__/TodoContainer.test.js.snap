// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoContainer renders correctly and fetches users and labels on mount 1`] = `
<div
  className="MuiPaper-root-1 MuiPaper-elevation2-5 MuiPaper-rounded-2"
>
  <h1
    className="MuiTypography-root-28 MuiTypography-display1-32 MuiTypography-gutterBottom-45 MuiTypography-alignLeft-40"
  >
    Loading...
  </h1>
</div>
`;
